content/documentation/pages/3-stream-developer-guides/3-programming-models/0-index.md

```diff
@@ -5,6 +5,196 @@ description: 'Programming models'
+```
+
+### Custom Processor
+
+For the data transformation between the source and sink steps, we will highlight a custom processor application and use that as a base to demonstrate different programming models.
+
+**Code:**
+
+<!--TABS-->
+<!--Message Channels-->
+
+```java
+@EnableBinding(Processor.class)
+public class SimpleStreamSampleProcessor {
+
+	@StreamListener(Processor.INPUT)
+	@SendTo(Processor.OUTPUT)
+	public String messenger(String data) {
+		return "Hello: " + data + "!";
+	}
+}
+```
+
+<!--Functional-->
+
+```java
+@EnableBinding(Processor.class)
+public class FunctionStreamSampleProcessor {
+
+	@Bean
+	public Function<String, String> messenger() {
+		return data -> "Hello: " + data + "!";
+	}
+}
+```
+
+<!--Kafka Streams-->
+
+```java
+@EnableBinding(KafkaStreamsProcessor.class)
+public class KafkaStreamsSampleProcessor {
+
+	@StreamListener("input")
+	@SendTo("output")
+	public KStream<String, String> messenger(KStream<String, String> data) {
+		return data.map((k, v) -> new KeyValue<>(null, "Hello: " + v + "!"));
+	}
+}
+```
+
+<!--Reactive-->
+
+```java
+@EnableBinding(Processor.class)
+public class ReactiveStreamSampleProcessor {
+
+	@StreamListener(Processor.INPUT)
+	@SendTo(Processor.OUTPUT)
+	public Flux<String> messenger(Flux<String> data) {
+		return data.map(incoming -> "Hello: " + incoming + "!");
+	}
+}
+```
+
+<!--END_TABS-->
+
+[[note]]
+| The business logic in the processor simply transforms the received payload by adding the "Hello: " prefix and then the "!" suffix in the end.
+|
+| The "same business logic" can be implemented with different programming models, and each of the variations is implementing a simple `messenger` function, which can be independently tested and evolved in isolation.
+|
+| **Takeaway**: Developers have the choice to choose from the available programming model styles.
+
+**Configuration: _(application.properties)_**
+
+<!--TABS-->
+<!--Message Channels-->
+
+```properties
+spring.cloud.stream.bindings.input.destination=incomingDataTopic
+spring.cloud.stream.bindings.output.destination=outgoingDataTopic
+```
+
+<!--Functional-->
+
+```properties
+spring.cloud.stream.bindings.input.destination=incomingDataTopic
+spring.cloud.stream.bindings.output.destination=outgoingDataTopic
+
+spring.cloud.stream.function.definition=messenger
+```
+
+<!--Kafka Streams-->
+
+```properties
+spring.cloud.stream.bindings.input.destination=incomingDataTopic
+spring.cloud.stream.bindings.output.destination=outgoingDataTopic
+
+spring.cloud.stream.kafka.streams.binder.applicationId=kstreams-sample
+```
+
+<!--Reactive-->
+
+```properties
+spring.cloud.stream.bindings.input.destination=incomingDataTopic
+spring.cloud.stream.bindings.output.destination=outgoingDataTopic
+```
+
+<!--END_TABS-->
+
+[[note]]
+| In Functional sample, it is _required_ to point to the `messenger` function using the `spring.cloud.stream.function.definition` property.
```

Comment:
In the latest `SCFn`, this property is no longer necessary unless the application has a composition model.

Reply:
Ah, good point. One less property to discuss. Done.